### PR TITLE
Add cost accounting for methods we missed

### PR DESCRIPTION
**Pull request by goral09**
## Overview
Some of the methods weren't charged for, this PR adds them and adjust cost of few others.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
Add link to corresponding JIRA issue.

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [ ] You assigned one person to review this PR

### If you are not a member of the core development team, please confirm:
- [ ] You signed the commit. Merging requires a signature. Please see the [developer wiki](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain) for instructions.
- [ ] Your GitHub account is also an account with [Travis CI](https://travis-ci.org). Unit tests will not run on your PR unless you have an account with Travis. Merge requires passed unit tests.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 1]
```diff
@@ -687,7 +689,7 @@ object Reduce {
                                       .map(
                                         keyValuePairs => GString(interpolate(lhs, keyValuePairs))
                                       )
-                           _ <- costAccountingAlg.charge(LOOKUP_COST * lhs.length)
+                           _ <- costAccountingAlg.charge(LOOKUP_COST * (lhs.length * rhs.size))
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: This type of change feels like it should have a test. Also the .length vs .size difference seems to stand out - there is no different right?

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Also why is this multiplicative versus the prepending costs additive?

Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Oh, it should be additive. It was late, thanks for catching that.

Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Because there's no `length` on `Map`.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala [hunk 2]
```diff
@@ -47,10 +47,14 @@ trait Costs {
   // + allocates byte array of the same size as `serializedSize`
   // + then it copies all elements of the Par
   def toByteArrayCost[T <: GeneratedMessage](a: T): Cost = Cost(a.serializedSize)
+  //TODO: adjust the cost of size method
+  def sizeMethodCost(size: Int): Cost = Cost(size % 100)
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: Why `% 100`? If we need a temporary implementation - why not just size?

Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: Cost equal to size of the collection is definitely too much. It is temporary but I felt like that big disproportion between simple `size` method and other things is just too much.

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: I'd feel more comfortable if somebody else chimed in here. To me, this should be just `size`, and even `size  / 100` doesn't seem safe to me.

It might be safe, but under the assumption that other parts of the cost model account for this cost. In such a case, I'd rather not be doing any reasoning about the cost model, because it can't be reasoned about locally.

To me, each part that incurs a computational / other cost should account for the cost itself. This also means that the least common denominator of all costs is the cost of the cheapest operation, and if e.g. reading from rspace shold be O(size) and size should be O(size), but the latter with much lower constant - then the rspace read should have a bigger constant.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala [hunk 3]
```diff
@@ -53,6 +53,10 @@ trait Costs {
   def toByteArrayCost[T <: GeneratedMessage](a: T): Cost = Cost(a.serializedSize)
   //TODO: adjust the cost of size method
   def sizeMethodCost(size: Int): Cost = Cost(size)
+  def sliceCost(from: Int, to: Int): Cost = {
+    val diff = to - from
+    if (diff <= 0) Cost(0) else Cost(diff)
+  }
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: I think the cost of slicing should be `from + (to - from) == to`, becuase we probably first need to pass `from` elements to then read the `to - from` of the ones we're interested in.

Also, we should probably not charge 100 for [1,2,3].slice(2, 100) (the cost should be parameterized - and capped by - collection size).

Did we do any range checking for costs so far? If not, I'd add it here and file a JIRA for adding it in other places.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 4]
```diff
@@ -955,7 +955,7 @@ object Reduce {
               ESetBody(base @ ParSet(basePs, _, _, _)),
               ESetBody(other @ ParSet(otherPs, _, _, _))
               ) =>
-            costAccountingAlg.charge(ADD_COST * otherPs.size) *> Applicative[M].pure[Expr](
+            costAccountingAlg.charge(unionCost(otherPs.size)) *> Applicative[M].pure[Expr](
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: :+1:

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SortSpec.scala [hunk 5]
```diff
@@ -14,7 +14,7 @@ class SortSpec extends FlatSpec with Matchers {
   "GroundSortMatcher" should "discern sets with and without remainder" in {
     assertOrder[Expr](
       ParSet(Seq.empty),
-      ParSet(Seq.empty, remainder = Some(FreeVar(0)))
+      ParSet(Seq.empty, connectiveUsed = false, remainder = Some(FreeVar(0)))
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/SortSpec.scala]: why this change? It doesn't help with the tests readability.

File: models/src/main/scala/coop/rchain/models/ParSet.scala [hunk 6]
```diff
@@ -36,13 +36,22 @@ object ParSet {
 
   def apply(
       ps: Seq[Par],
-      connectiveUsed: Boolean = false,
-      remainder: Option[Var] = None
+      connectiveUsed: Boolean,
+      remainder: Option[Var]
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/ParSet.scala]: May I suggest making `remainder` the 2nd argument on the occasion? :) Up to you, since it's an additional refactoring.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SortSpec.scala (hunk 5 above)
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/SortSpec.scala]: I think you can do `ParSet(Seq.empty, remainder = Some(FreeVar(0)))`

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala (hunk 3 above)
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: > Also, we should probably not charge 100 for [1,2,3].slice(2, 100) (the cost should be parameterized - and capped by - collection size).

good call.

> Did we do any range checking for costs so far? If not, I'd add it here and file a JIRA for adding it in other places.

I am writing tests at the very moment.

File: models/src/main/scala/coop/rchain/models/ParSet.scala (hunk 6 above)
Comment (goral09) [reply models/src/main/scala/coop/rchain/models/ParSet.scala]: It is 2nd argument.

File: models/src/main/scala/coop/rchain/models/ParSet.scala [hunk 7]
```diff
@@ -36,13 +36,23 @@ object ParSet {
...
+      remainder
+    )
   }
 
-  def updateLocallyFree(ps: SortedParHashSet): BitSet =
+  def apply(ps: Seq[Par]): ParSet =
+    apply(ps, None)
+
+  private def connectiveUsed(seq: Seq[Par]): Boolean =
+    seq.exists(_.connectiveUsed)
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/ParSet.scala]: We actually need the field, or we'll rely on our children's caches - which is a terrible thing for a parent to do

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala [hunk 8]
```diff
@@ -447,18 +446,15 @@ class VarMatcherSpec extends FlatSpec with Matchers with TimeLimits with TripleE
...
+      ParSet(Seq(GInt(2), GInt(5), EVar(FreeVar(0)), EVar(Wildcard(WildcardMsg()))))
     assertSpatialMatch(target, nonMatchingPattern, None)
   }
 
   it should "work with wildcard remainders" in {
     val targetElements = Seq[Par](GInt(1), GInt(2), GInt(3), GInt(4), GInt(5))
     val target: Expr   = ParSet(targetElements)
     val pattern: Expr =
-      ParSet(Seq(GInt(1), GInt(4)), connectiveUsed = true, remainder = Wildcard(WildcardMsg()))
+      ParSet(Seq(GInt(1), GInt(4)), remainder = Wildcard(WildcardMsg()))
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: ~I think this will no longer have `connectiveUsed` in your implementeation.~

It will. But we should add a test if no test of ParSet fails when you don't include remainder.

We should add a generative test of the form:
```scala
forAll { (a: Expr, b: Expr, r: Var) =>
   ParSet(Seq(a, b), remainer = r).connectiveUsed == Seq(a, b, r).conains(_.connectiveUsed) //assumin Var has connectiveUsed == true...
}
```
or sth like that

File: models/src/main/scala/coop/rchain/models/ParSet.scala (hunk 7 above)
Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/ParSet.scala]: Also, we should probably make the `connectiveUsed` computation lazy... At this point I think we should maybe allow passing it from outside as it used to be?

Comment (goral09) [reply models/src/main/scala/coop/rchain/models/ParSet.scala]: > We actually need the field, or we'll rely on our children's caches - which is a terrible thing for a parent to do

what caches? I think this version is safer than it was before.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala (hunk 8 above)
Comment (goral09) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: There is a test for that: https://github.com/rchain/rchain/blob/72c6dff860ee6147443feac60b5fddf8367d058f/rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala#L434

File: models/src/main/scala/coop/rchain/models/ParSet.scala (hunk 7 above)
Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/ParSet.scala]: It is and I'd very much welcome the change. `connectiveUsed` exists on the tree so that it doesn't have to be re-evaluated each time we ask about it. Ohterwise we'd just use the HasLocallyFree typeclass.
